### DISCUSSION TITLE
1.2 coming soon, what will change

### DISCUSSION THREAD

**Discussion by kagenash1** [kagenash1/godot-behavior-tree#17] [Announcements]
(The previous 1.1.3 version is now 1.2. The previous 1.2 is now 2.0.)

The upcoming version 1.2 will bring several fixes and optimizations as usual. There are a couple new features and a new type of decorator.

Changes:
- Removed the on_tick() function.
- Added pre_tick() and post_tick() functions which get called before and after tick() (useful in decorators)
- Added abort_tree() signal on BTNode to abort the tree execution after a tick() call
- BTRepeat and BTRepeatUntil now have a frequency property to specify the rate of tick() during repetition
- BTWait now allows to wait for a time specified in a blackboard entry.
- Several refactoring based on community feedback to make the code more readable
- Improved script templates 

New nodes:
- **BTConditional** node, inheriting from BTDecorator. From now on, conditions are carried out through these nodes and not with decorators anymore. This means your old conditionals will still work, but it's encouraged that they get refactored to BTConditional if you upgrade to the 1.2 version. The script template explains how.

Maybe there will be more but these are the news so far.

**Comment by MadFlyFish** [comment]
Great ！It's getting better and better. Can't wait to see the new version！

**Comment by kagenash1** [comment]
To update on the previous post, 1.2 will probably come out after Godot 3.3, at this point, so we can be sure it works with the new version.
So far optimization has continued on several fronts.

**Reply by MadFlyFish** [reply]
Glad to hear that. Keep punching my friend.
Recently, I have been busy with company projects. Over the past two days, I started to work on the Btree internal editor.
However It proved a bit complicated to make use of GraphEdit. :)

**Reply by kagenash1** [reply]
Great! I'm glad you're willing to contribute. Some of the changes are up in the 1.2 branch.

**Reply by MadFlyFish** [reply]
Good news my friend.
Godot 3.3 has come out today. ：)

**Reply by kagenash1** [reply]
Yes, I'm already testing it and adding the final touches. It's coming out today or tomorrow (or Toyota).

**Reply by MadFlyFish** [reply]
Hey，how is everything going?
Yesterday I tried XSM according to your advice. 
However it turns out behavior tree is easier to use when it comes to AI :)
And state switching and flow interrupting are quite necessary.

**Reply by ashfidable** [reply]
Waiting for 1.2! I do agree with state interrupting too. I am mixing btree and fsm together. I got a good flow but its bit hacky to change from one btree to another sadly.

**Reply by MadFlyFish** [reply]
> Waiting for 1.2! I do agree with state interrupting too. I am mixing btree and fsm together. I got a good flow but its bit hacky to change from one btree to another sadly.

Great. But I am still quite confused about how to mixing them together.

**Comment by kagenash1** [comment]
Sorry for the long wait! The new version is out. Please test it out so I can make a new release! 

I suggest you to backup your project, because you need to tweak things a bit and double check your trees in order to update because it's not 100% backwards compatible.

**Reply by MadFlyFish** [reply]
Glad to see the new beta finally
I will test it out right now. : )
Thanks for your effort.
